File: app/src/main/java/me/arianb/usb_hid_client/input_views/TouchpadView.java
package me.arianb.usb_hid_client.input_views;

import android.annotation.SuppressLint;
import android.content.Context;
import android.util.AttributeSet;
import android.view.MotionEvent;
import android.view.VelocityTracker;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import me.arianb.usb_hid_client.report_senders.MouseSender;
import timber.log.Timber;

// TODO: address the linting issue below
@SuppressLint("ClickableViewAccessibility")
public class TouchpadView extends androidx.appcompat.widget.AppCompatTextView {
    private VelocityTracker mVelocityTracker = null;
    private static final float DEADZONE = 0.3F;

    // Vars for tracking a single touch event, which I'm defining as: ACTION_DOWN, (anything), ACTION_UP
    private int currentPointerCount;

    public TouchpadView(@NonNull Context context) {
        super(context);
    }

    public TouchpadView(@NonNull Context context, @Nullable AttributeSet attrs) {
        super(context, attrs);
    }

    public TouchpadView(@NonNull Context context, @Nullable AttributeSet attrs, int defStyleAttr) {
        super(context, attrs, defStyleAttr);
    }

    public void setTouchListeners(MouseSender mouseSender) {
        // TODO:
        //  - add gestures
        //      - on double click and drag, send report without "release" until sending release on finger up
        //      - on long press and drag, send report without "release" until sending release on finger up
        //      - on two finger scroll, send scroll events
        this.setOnTouchListener((view, motionEvent) -> {
            int action = motionEvent.getActionMasked();
            int index = motionEvent.getActionIndex();
            int pointerId = motionEvent.getPointerId(index);

            //Timber.d("motionEvent %d (x, y): (%f, %f)", action, motionEvent.getX(), motionEvent.getY());
            //Timber.d("current pointer count: %d (pressure: %s)", currentPointerCount, motionEvent.getPressure());

            switch (action) {
                case MotionEvent.ACTION_DOWN:
                    currentPointerCount = 1;
                    Timber.d("Action Down");

                    if (mVelocityTracker == null) {
                        mVelocityTracker = VelocityTracker.obtain();
                    } else {
                        mVelocityTracker.clear();
                    }
                    mVelocityTracker.addMovement(motionEvent);
                    break;
                case MotionEvent.ACTION_MOVE:
                    //Timber.d("Action Move");
                    mVelocityTracker.addMovement(motionEvent);

                    // Compute velocity (cap it to byte because the report uses a byte per axis)
                    mVelocityTracker.computeCurrentVelocity(10, Byte.MAX_VALUE);
                    float xVelocity = mVelocityTracker.getXVelocity(pointerId);
                    float yVelocity = mVelocityTracker.getYVelocity(pointerId);

                    // Scale up velocities < 1 in magnitude (accounting for deadzone) to allow for precise movements
                    xVelocity = scaleWithDeadzone(xVelocity);
                    yVelocity = scaleWithDeadzone(yVelocity);

                    byte x = (byte) xVelocity;
                    byte y = (byte) yVelocity;

                    mouseSender.move(x, y);
                    break;
                case MotionEvent.ACTION_UP:
                    Timber.d("Action Up (max pointer count = %d)", currentPointerCount);

                    final long pointerDownTimeMillis = motionEvent.getEventTime() - motionEvent.getDownTime();
                    Timber.d("Pointer was down for %d milliseconds", pointerDownTimeMillis);

                    // If user has been holding down for a while, don't send any clicks, they're probably dragging
                    if (pointerDownTimeMillis > 300) {
                        break;
                    }

                    // TODO: if finger has traveled too far, don't send any click? I already have the timeout thing though.

                    switch (currentPointerCount) {
                        case 1:
                            mouseSender.click(MouseSender.MOUSE_BUTTON_LEFT);
                            break;
                        case 2:
                            mouseSender.click(MouseSender.MOUSE_BUTTON_RIGHT);
                            break;
                        case 3:
                            mouseSender.click(MouseSender.MOUSE_BUTTON_MIDDLE);
                            break;
                    }

                    currentPointerCount = 0;
                    break;
                case MotionEvent.ACTION_POINTER_DOWN:
                    currentPointerCount++;
                    //Timber.d("Action Pointer Down (pointer count = %s)", motionEvent.getPointerCount());

                    break;
                case MotionEvent.ACTION_POINTER_UP:
                    Timber.d("Action Pointer Up");

                    // this pointer decrement is conceptually happening, but I'm not doing it because I want ACTION_UP
                    // to use the max pointer count.
                    // currentPointerCount--;
                    break;
                case MotionEvent.ACTION_CANCEL:
                    Timber.d("Action Cancel");
                    mVelocityTracker.recycle(); // Return a VelocityTracker object back to be re-used by others.
                    currentPointerCount = 0;
                    break;
            }
            return true;

        });
    }

    private float scaleWithDeadzone(float inputVelocity) {
        if (Math.abs(inputVelocity) != 0 && Math.abs(inputVelocity) > DEADZONE) {
            if (inputVelocity < 0) {
                return (float) Math.floor(inputVelocity);
            } else {
                return (float) Math.ceil(inputVelocity);
            }
        } else {
            return inputVelocity;
        }
    }
}

File: app/src/main/java/me/arianb/usb_hid_client/input_views/TouchpadView.kt
package me.arianb.usb_hid_client.input_views

import android.annotation.SuppressLint
import android.content.Context
import android.os.Build
import android.util.AttributeSet
import android.view.Gravity
import android.view.InputDevice
import android.view.MotionEvent
import android.view.View
import androidx.appcompat.widget.AppCompatTextView
import androidx.compose.foundation.BorderStroke
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Surface
import androidx.compose.runtime.Composable
import androidx.compose.runtime.collectAsState
import androidx.compose.runtime.getValue
import androidx.compose.ui.Modifier
import androidx.compose.ui.res.stringResource
import androidx.compose.ui.unit.dp
import androidx.compose.ui.viewinterop.AndroidView
import androidx.lifecycle.viewmodel.compose.viewModel
import me.arianb.usb_hid_client.MainViewModel
import me.arianb.usb_hid_client.R
import me.arianb.usb_hid_client.report_senders.PointerDeviceSender
import me.arianb.usb_hid_client.ui.utils.getColorByTheme
import timber.log.Timber

// LEGACY: migrate this to Compose

// From my understanding of this lint warning, I don't think it applies here.
@SuppressLint("ClickableViewAccessibility")
class TouchpadView : AppCompatTextView {
    private var currentScanTime: UShort = getScanTime()

    constructor(context: Context) : super(context)
    constructor(context: Context, attrs: AttributeSet?) : super(context, attrs)
    constructor(context: Context, attrs: AttributeSet?, defStyleAttr: Int) : super(
        context,
        attrs,
        defStyleAttr
    )

    fun setTouchListeners(touchpadSender: PointerDeviceSender) {
        setOnTouchListener { _: View?, motionEvent: MotionEvent ->
            val (pointerID, pointerX, pointerY) = getPointerTriple(motionEvent, pointerIndex = motionEvent.actionIndex)

            // Scan time is reset when pointer 0 is sent
            if (pointerID == 0) {
                currentScanTime = getScanTime()
            }

            val pointerCount = motionEvent.pointerCount
            when (val action = motionEvent.actionMasked) {
                MotionEvent.ACTION_DOWN -> {
                    Timber.v("Action Down")
                    touchpadSender.send(pointerID, true, pointerX, pointerY, currentScanTime, pointerCount)
                }

                MotionEvent.ACTION_POINTER_DOWN -> {
                    Timber.v("Action Pointer Down")
                    touchpadSender.send(pointerID, true, pointerX, pointerY, currentScanTime, pointerCount)
                }

                MotionEvent.ACTION_MOVE -> {
                    Timber.v("Action Move")
                    for (index in 0..<pointerCount) {
                        val (thisID, thisX, thisY) = getPointerTriple(motionEvent, index)

                        touchpadSender.send(thisID, true, thisX, thisY, currentScanTime, pointerCount)
                    }
                }

                MotionEvent.ACTION_UP -> {
                    Timber.v("Action Up")
                    touchpadSender.send(pointerID, false, pointerX, pointerY, currentScanTime, pointerCount)
                }

                MotionEvent.ACTION_POINTER_UP -> {
                    Timber.v("Action Pointer Up")
                    touchpadSender.send(pointerID, false, pointerX, pointerY, currentScanTime, pointerCount)
                }

                MotionEvent.ACTION_CANCEL -> {
                    Timber.v("Action Cancel")
                    touchpadSender.send(pointerID, false, pointerX, pointerY, currentScanTime, pointerCount)
                }

                else -> {
                    Timber.w("UNHANDLED ACTION CONSTANT: %s", action)
                }
            }
            true
        }
    }
}

private fun getPointerTriple(motionEvent: MotionEvent, pointerIndex: Int): Triple<Int, Int, Int> {
    val pointerID = motionEvent.getPointerId(pointerIndex)

    val (rawPointerX, rawPointerY) = if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.Q) {
        Pair(motionEvent.getRawX(pointerIndex), motionEvent.getRawY(pointerIndex))
    } else {
        Pair(motionEvent.getX(pointerIndex), motionEvent.getY(pointerIndex))
    }

    // NOTE: MotionEvent has a bunch of properties (and properties of those properties) which are platform types.
    //       I'm writing this note to be explicit about the fact that the following code should be treated cautiously so
    //       as to not cause NPEs.
    // --- start of unsafe code ---
    val device: InputDevice? = motionEvent.device

    // If null, just use some hardcoded safe-ish values
    val xMax: Float = device?.getMotionRange(MotionEvent.AXIS_X)?.max ?: 1500f
    val yMax: Float = device?.getMotionRange(MotionEvent.AXIS_Y)?.max ?: 3000f
    // --- end of unsafe code ---

    // Get device rotation because if the device is suddenly a wide rectangle instead of a tall rectangle, then the
    // math changes.
    val isRotated = motionEvent.orientation != 0f

    val (pointerX, pointerY) = adjustRange(
        point = Pair(rawPointerX.toInt(), rawPointerY.toInt()),
        max = Pair(xMax, yMax),
        isRotated
    )

    return Triple(pointerID, pointerX, pointerY)
}

// "Stretches" the values of the points to use up the entire logical range.
private fun adjustRange(point: Pair<Int, Int>, max: Pair<Float, Float>, isRotated: Boolean): Pair<Int, Int> {
    Timber.d("DEVICE COORDINATE MAX = (%f, %f)", max.first, max.second)

    val (logicalMaxX, logicalMaxY) = if (isRotated) {
        // This works, but I'm not sure if it's okay to just be sending values higher than the logical maximum
        Pair(5000, 2500)
    } else {
        Pair(2500, 5000)
    }

    val (pointerMaxX, pointerMaxY) = if (isRotated) {
        Pair(max.second, max.first)
    } else {
        max
    }

    val xRatio: Float = logicalMaxX / pointerMaxX
    val yRatio: Float = logicalMaxY / pointerMaxY

    val adjustedX = (point.first * xRatio).toInt()
    val adjustedY = (point.second * yRatio).toInt()

    // This will probably never actually be necessary, but might as well do it just in case.
    val finalX = adjustedX.coerceIn(0, logicalMaxX)
    val finalY = adjustedY.coerceIn(0, logicalMaxY)

    return Pair(finalX, finalY)
}

fun getScanTime(): UShort {
    // Convert nanoseconds to microseconds
    val microTime = System.nanoTime() / 1000

    // Convert microseconds to 100s of microseconds
    val hundredMicroTime = microTime / 100

    return hundredMicroTime.toUShort()
}

/**
 * Helper function to convert between types.
 */
fun PointerDeviceSender.send(
    pointerID: Int,
    tipSwitch: Boolean,
    x: Int,
    y: Int,
    currentScanTime: UShort,
    pointerCount: Int
) = send(
    pointerID.toByte(),
    tipSwitch,
    x.toShort(),
    y.toShort(),
    currentScanTime,
    pointerCount.toByte()
)

@Composable
fun Touchpad(mainViewModel: MainViewModel = viewModel()) {
    val touchpadText = stringResource(R.string.touchpad_label)
    val touchpadSender by mainViewModel.touchpadSender.collectAsState()

    val textColor = getColorByTheme()

    Surface(
        modifier = Modifier
//            .pointerInput(Unit) {
//
//            }
            .fillMaxSize(),
        color = MaterialTheme.colorScheme.background,
        border = BorderStroke(2.dp, MaterialTheme.colorScheme.primary)
    ) {
        // For when I want to Compose-ify this
//        Text(
//            text = stringResource(R.string.touchpad_label),
//            modifier = Modifier.wrapContentHeight(Alignment.CenterVertically),
//            textAlign = TextAlign.Center
//        )
        AndroidView(
            factory = { context ->
                TouchpadView(context).apply {
                    text = touchpadText
                    textSize = 22f
                    gravity = Gravity.CENTER
                    setTouchListeners(touchpadSender)
                }
            },
            update = {
                if (textColor != null) {
                    it.setTextColor(textColor)
                }
            }
        )
    }
}
